Using an Enum in ConfigProperties doesn't work in native mode

**Steps to reproduce**

* Checkout: https://github.com/dejanb/quarkus-quickstarts/tree/enum-reproducer/config-quickstart
* `mvn clean package -DskipTests -Dnative`
* `curl http://0.0.0.0:8080/enum/`

Error is:

```
Caused by: java.lang.IllegalArgumentException: SRCFG00013: No Converter registered for class org.acme.config.NumberEnum
```

cc @dejanb

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ClassConfigPropertiesUtil.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ClassConfigPropertiesUtil.java
@@ -34,6 +34,7 @@ import io.quarkus.arc.deployment.ConfigPropertyBuildItem;
 import io.quarkus.arc.deployment.configproperties.ConfigPropertiesUtil.ReadOptionalResponse;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.bean.JavaBeanUtil;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveMethodBuildItem;
 import io.quarkus.deployment.util.HashUtil;
 import io.quarkus.gizmo.BranchResult;
 import io.quarkus.gizmo.BytecodeCreator;
@@ -125,6 +126,7 @@ final class ClassConfigPropertiesUtil {
             ClassCreator producerClassCreator, String prefixStr, ConfigProperties.NamingStrategy namingStrategy,
             boolean failOnMismatchingMember,
             boolean needsQualifier, IndexView applicationIndex,
+            BuildProducer<ReflectiveMethodBuildItem> reflectiveMethods,
             BuildProducer<ConfigPropertyBuildItem> configProperties) {
 
         if (!configPropertiesClassInfo.hasNoArgsConstructor()) {
@@ -178,7 +180,7 @@ final class ClassConfigPropertiesUtil {
             }
 
             ResultHandle configObject = populateConfigObject(classLoader, configPropertiesClassInfo, prefixStr, namingStrategy,
-                    failOnMismatchingMember, methodCreator, applicationIndex, configProperties);
+                    failOnMismatchingMember, methodCreator, applicationIndex, reflectiveMethods, configProperties);
 
             if (needsValidation) {
                 createValidationCodePath(methodCreator, configObject, prefixStr);
@@ -210,6 +212,7 @@ final class ClassConfigPropertiesUtil {
     private static ResultHandle populateConfigObject(ClassLoader classLoader, ClassInfo configClassInfo, String prefixStr,
             ConfigProperties.NamingStrategy namingStrategy, boolean failOnMismatchingMember, MethodCreator methodCreator,
             IndexView applicationIndex,
+            BuildProducer<ReflectiveMethodBuildItem> reflectiveMethods,
             BuildProducer<ConfigPropertyBuildItem> configProperties) {
         String configObjectClassStr = configClassInfo.name().toString();
         ResultHandle configObject = methodCreator.newInstance(MethodDescriptor.ofConstructor(configObjectClassStr));
@@ -276,6 +279,22 @@ final class ClassConfigPropertiesUtil {
                         populateTypicalProperty(methodCreator, configObject, configPropertyBuildItemCandidates,
                                 currentClassInHierarchy, field, useFieldAccess, fieldType, setter, mpConfig,
                                 getFullConfigName(prefixStr, namingStrategy, field));
+
+                        // we need to register the 'valueOf' method of the enum for reflection because
+                        // that is the method that SmallryeConfig uses for conversion of enums
+                        List<MethodInfo> methods = fieldTypeClassInfo.methods();
+                        for (MethodInfo method : methods) {
+                            if (!method.name().equals("valueOf")) {
+                                continue;
+                            }
+                            if (method.parameters().size() != 1) {
+                                continue;
+                            }
+                            if (method.parameters().get(0).name().equals(DotNames.STRING)) {
+                                reflectiveMethods.produce(new ReflectiveMethodBuildItem(method));
+                                break;
+                            }
+                        }
                     } else {
                         if (!fieldTypeClassInfo.hasNoArgsConstructor()) {
                             throw new IllegalArgumentException(
@@ -291,7 +310,7 @@ final class ClassConfigPropertiesUtil {
                         ResultHandle nestedConfigObject = populateConfigObject(classLoader, fieldTypeClassInfo,
                                 getFullConfigName(prefixStr, namingStrategy, field), namingStrategy, failOnMismatchingMember,
                                 methodCreator,
-                                applicationIndex, configProperties);
+                                applicationIndex, reflectiveMethods, configProperties);
                         createWriteValue(methodCreator, configObject, field, setter, useFieldAccess, nestedConfigObject);
                     }
                 } else {

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertiesBuildStep.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertiesBuildStep.java
@@ -27,6 +27,7 @@ import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveMethodBuildItem;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.ClassOutput;
 
@@ -103,6 +104,7 @@ public class ConfigPropertiesBuildStep {
             BuildProducer<GeneratedClassBuildItem> generatedClasses,
             BuildProducer<GeneratedBeanBuildItem> generatedBeans,
             BuildProducer<RunTimeConfigurationDefaultBuildItem> defaultConfigValues,
+            BuildProducer<ReflectiveMethodBuildItem> reflectiveMethods,
             BuildProducer<ConfigPropertyBuildItem> configProperties) {
         if (configPropertiesMetadataList.isEmpty()) {
             return;
@@ -148,7 +150,7 @@ public class ConfigPropertiesBuildStep {
                         Thread.currentThread().getContextClassLoader(), classInfo, producerClassCreator,
                         configPropertiesMetadata.getPrefix(), configPropertiesMetadata.getNamingStrategy(),
                         configPropertiesMetadata.isFailOnMismatchingMember(), configPropertiesMetadata.isNeedsQualifier(),
-                        combinedIndex.getIndex(), configProperties);
+                        combinedIndex.getIndex(), reflectiveMethods, configProperties);
                 if (needsValidation) {
                     configClassesThatNeedValidation.add(classInfo.name());
                 }

--- a/integration-tests/main/src/main/java/io/quarkus/it/config/ConfigPropertiesResource.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/config/ConfigPropertiesResource.java
@@ -19,7 +19,7 @@ public class ConfigPropertiesResource {
 
     @GET
     public String greet() {
-        return greetingConfiguration.message + greetingConfiguration.suffix;
+        return greetingConfiguration.message + greetingConfiguration.number + greetingConfiguration.suffix;
     }
 
     @ConfigProperties(prefix = "configproperties")
@@ -28,5 +28,11 @@ public class ConfigPropertiesResource {
         public String message;
         public String suffix = "!";
         public BigDecimal other;
+        public NumberEnum number;
+    }
+
+    public enum NumberEnum {
+        ONE,
+        TWO;
     }
 }

--- a/integration-tests/main/src/main/resources/application.properties
+++ b/integration-tests/main/src/main/resources/application.properties
@@ -36,6 +36,7 @@ microprofile.custom.value = 456
 microprofile.cidr-address=10.0.0.20/24
 configproperties.message=Hello
 configproperties.other=1
+configproperties.number=ONE
 
 quarkus.swagger-ui.always-include=true
 

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/ConfigPropertiesTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/ConfigPropertiesTestCase.java
@@ -14,6 +14,6 @@ public class ConfigPropertiesTestCase {
     public void testConfigPropertiesProperlyInjected() {
         RestAssured
                 .when().get("/configuration-properties")
-                .then().body(is("Hello!"));
+                .then().body(is("HelloONE!"));
     }
 }

